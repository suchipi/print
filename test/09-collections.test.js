const { print } = require("./helpers");

describe("Collections", () => {
	const A = {
		colour: 0x00ff00,
		width: 28.52,
		height: 10.2,
		range: [-20, 12],
	};
	const B = {
		name: "John",
		age: "Older than I look",
		occupation: "Larrikin",
		country: "Australia",
		city: "Melbourne",
	};

	describe("Maps", () => {
		it("prints string-type entries", () => {
			expect(
				print(
					new Map([
						["alpha", "A"],
						["beta", "B"],
						["gamma", "G"],
						["delta", "D"],
					])
				)
			).toMatchSnapshot();
		});

		it("prints symbol-type entries", () => {
			const a = Symbol("Alpha");
			expect(print(new Map([["A", a]]))).toMatchSnapshot();
			expect(print(new Map([[a, null]]))).toMatchSnapshot();
			expect(print(new Map([[a, "A"]]))).toMatchSnapshot();
			expect(print(new Map([[null, a]]))).toMatchSnapshot();
			expect(print(new Map([[Symbol.iterator, a]]))).toMatchSnapshot();
			expect(print(new Map([[a, Symbol.iterator]]))).toMatchSnapshot();
		});

		it("prints object-type keys", () => {
			expect(
				print(
					new Map([
						[{ a: "a", A: "A" }, "alpha"],
						[{ b: "b", B: "B" }, "beta"],
						[{ g: "g", G: "G" }, "gamma"],
						[{ d: "d", D: "D" }, "delta"],
					])
				)
			).toMatchSnapshot();
			expect(
				print(
					new Map([
						[A, "A"],
						[B, "B"],
					])
				)
			).toMatchSnapshot();
			expect(print(new Map([[{ A, B }, "AB"]]))).toMatchSnapshot();
		});

		it("prints object-type values", () => {
			expect(
				print(
					new Map([
						["alpha", { a: "a", A: "A" }],
						["beta", { b: "b", B: "B" }],
						["gamma", { g: "g", G: "G" }],
						["delta", { d: "d", D: "D" }],
					])
				)
			).toMatchSnapshot();
			expect(
				print(
					new Map([
						[
							"alphabeta",
							{
								alpha: { a: "a", A: "A" },
								beta: { b: "b", B: "B" },
							},
						],
					])
				)
			).toMatchSnapshot();
		});

		it("prints properties", () => {
			const map = new Map([
				["A", "a"],
				["B", "b"],
				["C", "c"],
			]);
			map.name = "Quxabaz";
			map.customProperty = {
				foo: "Bar",
				baz: "Quux",
			};
			expect(print(map)).toMatchSnapshot();
		});

		it("indicates references using `->`", () => {
			const list = [1];
			const mappedKey = new Map([[list, "list"]]);
			const mappedValue = new Map([["list", list]]);
			expect(print([list, mappedKey], "input")).toMatchSnapshot();
			expect(print([list, mappedValue], "input")).toMatchSnapshot();
			list.pop();
			expect(print([list, mappedKey, mappedValue])).toMatchSnapshot();
		});

		it("prints empty maps on one line", () => {
			const map = new Map();
			expect(print(map)).toMatchSnapshot();
			map.foo = "Bar";
			expect(print(map)).toMatchSnapshot();
		});

		it("catches errors thrown by faulty iterators", () => {
			class BadMap extends Map {
				[Symbol.iterator]() {
					let count = 0;
					return {
						next() {
							if (++count >= 3) {
								const error = new RangeError("You're out");
								error.strikes = 3;
								throw error;
							}
							return { value: [count, count] };
						},
					};
				}
			}
			expect(print(new BadMap())).toMatchSnapshot();
		});
	});

	describe("Sets", () => {
		it("prints basic entries", () => {
			const set = new Set(["A", "B", 0xcc]);
			expect(print(set)).toMatchSnapshot();
		});

		it("prints symbol-type entries", () => {
			const set = new Set([Symbol.iterator, Symbol("Foo")]);
			expect(print(set)).toMatchSnapshot();
		});

		it("prints object-type entries", () => {
			const set = new Set(["0", A, B, "C"]);
			expect(print(set)).toMatchSnapshot();
		});

		it("prints nested sets", () => {
			const nest = new Set([1, A, "3", new Set([2, B, "4"])]);
			expect(print(nest)).toMatchSnapshot();
		});

		it("prints properties", () => {
			const input = new Set(["A", "B", "C"]);
			input.name = "Quxabaz";
			input.customProperty = {
				foo: "Bar",
				baz: "Quux",
			};
			expect(print(input)).toMatchSnapshot();
		});

		it("indicates references using `->`", () => {
			const bar = { name: "Bar" };
			const input = { foo: "Foo", bar, baz: new Set([bar]) };
			expect(print(input)).toMatchSnapshot();
		});

		it("disambiguates entries and numeric properties", () => {
			const a = { b: "c" };
			const set = new Set([a]);
			set.foo = { bar: "Baz" };
			set.add(set.foo);
			set[1] = { nah: "Nope" };
			set.bar = set[1];
			expect(print({ set }, "input")).toMatchSnapshot();
		});

		it("prints empty sets on one line", () => {
			const set = new Set();
			expect(print(set)).toMatchSnapshot();
			set.foo = "Bar";
			expect(print(set)).toMatchSnapshot();
		});

		it("catches errors thrown by faulty iterators", () => {
			class BadSet extends Set {
				[Symbol.iterator]() {
					let count = 0;
					return {
						next() {
							if (++count >= 3) {
								const error = new RangeError("You're out");
								error.strikes = 3;
								throw error;
							}
							return { value: count };
						},
					};
				}
			}
			expect(print(new BadSet())).toMatchSnapshot();
		});
	});

	describe("WeakMaps", () => {
		it("prints properties", () => {
			const input = new WeakMap();
			input.a = "ABC";
			expect(print(input)).toMatchSnapshot();
			input.b = "XYZ";
			expect(print(input)).toMatchSnapshot();
		});

		it("prints empty maps on one line", () => {
			expect(print(new WeakMap())).toMatchSnapshot();
		});
	});

	describe("WeakSets", () => {
		it("prints properties", () => {
			const input = new WeakSet();
			input.a = "ABC";
			expect(print(input)).toMatchSnapshot();
			input.b = "XYZ";
			expect(print(input)).toMatchSnapshot();
		});

		it("prints empty sets on one line", () => {
			expect(print(new WeakSet())).toMatchSnapshot();
		});
	});
});
