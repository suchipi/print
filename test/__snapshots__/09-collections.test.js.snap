// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collections Maps catches errors thrown by faulty iterators 1`] = `
BadMap {
	0.key => 1
	0.value => 1
	
	1.key => 2
	1.value => 2
	
	RangeError {
		RangeError: You're out
			at Object.next (<rootDir>/test/09-collections.test.js:LINE:COL)
			at print (<rootDir>/print.js:LINE:COL)
			at realPrint (<rootDir>/test/helpers.js:LINE:COL)
			at Object.print (<rootDir>/test/09-collections.test.js:LINE:COL)
			at Promise.then.completed (<rootDir>/node_modules/jest-circus/build/utils.js:LINE:COL)
			at new Promise (<anonymous>)
			at callAsyncCircusFn (<rootDir>/node_modules/jest-circus/build/utils.js:LINE:COL)
			at _callCircusTest (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
			at _runTest (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
			at _runTestsForDescribeBlock (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
		
		strikes: 3
	}
}
`;

exports[`Collections Maps indicates references using \`->\` 1`] = `
input: [
	[
		1
	]
	Map {
		0.key -> input[0]
		0.value => "list"
	}
]
`;

exports[`Collections Maps indicates references using \`->\` 2`] = `
input: [
	[
		1
	]
	Map {
		0.key => "list"
		0.value -> input[0]
	}
]
`;

exports[`Collections Maps indicates references using \`->\` 3`] = `
[
	[]
	Map {
		0.key -> {root}[0]
		0.value => "list"
	}
	Map {
		0.key => "list"
		0.value -> {root}[0]
	}
]
`;

exports[`Collections Maps prints empty maps on one line 1`] = `Map {}`;

exports[`Collections Maps prints empty maps on one line 2`] = `
Map {
	foo: "Bar"
}
`;

exports[`Collections Maps prints object-type keys 1`] = `
Map {
	0.key => {
		a: "a"
		A: "A"
	}
	0.value => "alpha"
	
	1.key => {
		b: "b"
		B: "B"
	}
	1.value => "beta"
	
	2.key => {
		g: "g"
		G: "G"
	}
	2.value => "gamma"
	
	3.key => {
		d: "d"
		D: "D"
	}
	3.value => "delta"
}
`;

exports[`Collections Maps prints object-type keys 2`] = `
Map {
	0.key => {
		colour: 65280
		width: 28.52
		height: 10.2
		range: [
			-20
			12
		]
	}
	0.value => "A"
	
	1.key => {
		name: "John"
		age: "Older than I look"
		occupation: "Larrikin"
		country: "Australia"
		city: "Melbourne"
	}
	1.value => "B"
}
`;

exports[`Collections Maps prints object-type keys 3`] = `
Map {
	0.key => {
		A: {
			colour: 65280
			width: 28.52
			height: 10.2
			range: [
				-20
				12
			]
		}
		B: {
			name: "John"
			age: "Older than I look"
			occupation: "Larrikin"
			country: "Australia"
			city: "Melbourne"
		}
	}
	0.value => "AB"
}
`;

exports[`Collections Maps prints object-type values 1`] = `
Map {
	0.key => "alpha"
	0.value => {
		a: "a"
		A: "A"
	}
	
	1.key => "beta"
	1.value => {
		b: "b"
		B: "B"
	}
	
	2.key => "gamma"
	2.value => {
		g: "g"
		G: "G"
	}
	
	3.key => "delta"
	3.value => {
		d: "d"
		D: "D"
	}
}
`;

exports[`Collections Maps prints object-type values 2`] = `
Map {
	0.key => "alphabeta"
	0.value => {
		alpha: {
			a: "a"
			A: "A"
		}
		beta: {
			b: "b"
			B: "B"
		}
	}
}
`;

exports[`Collections Maps prints properties 1`] = `
Map {
	0.key => "A"
	0.value => "a"
	
	1.key => "B"
	1.value => "b"
	
	2.key => "C"
	2.value => "c"
	
	name: "Quxabaz"
	customProperty: {
		foo: "Bar"
		baz: "Quux"
	}
}
`;

exports[`Collections Maps prints string-type entries 1`] = `
Map {
	0.key => "alpha"
	0.value => "A"
	
	1.key => "beta"
	1.value => "B"
	
	2.key => "gamma"
	2.value => "G"
	
	3.key => "delta"
	3.value => "D"
}
`;

exports[`Collections Maps prints symbol-type entries 1`] = `
Map {
	0.key => "A"
	0.value => Symbol(Alpha)
}
`;

exports[`Collections Maps prints symbol-type entries 2`] = `
Map {
	0.key => Symbol(Alpha)
	0.value => null
}
`;

exports[`Collections Maps prints symbol-type entries 3`] = `
Map {
	0.key => Symbol(Alpha)
	0.value => "A"
}
`;

exports[`Collections Maps prints symbol-type entries 4`] = `
Map {
	0.key => null
	0.value => Symbol(Alpha)
}
`;

exports[`Collections Maps prints symbol-type entries 5`] = `
Map {
	0.key => @@iterator
	0.value => Symbol(Alpha)
}
`;

exports[`Collections Maps prints symbol-type entries 6`] = `
Map {
	0.key => Symbol(Alpha)
	0.value => @@iterator
}
`;

exports[`Collections Sets catches errors thrown by faulty iterators 1`] = `
BadSet {
	0 => 1
	1 => 2
	RangeError {
		RangeError: You're out
			at Object.next (<rootDir>/test/09-collections.test.js:LINE:COL)
			at v (<rootDir>/print.js:LINE:COL)
			at realPrint (<rootDir>/test/helpers.js:LINE:COL)
			at Object.print (<rootDir>/test/09-collections.test.js:LINE:COL)
			at Promise.then.completed (<rootDir>/node_modules/jest-circus/build/utils.js:LINE:COL)
			at new Promise (<anonymous>)
			at callAsyncCircusFn (<rootDir>/node_modules/jest-circus/build/utils.js:LINE:COL)
			at _callCircusTest (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
			at _runTest (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
			at _runTestsForDescribeBlock (<rootDir>/node_modules/jest-circus/build/run.js:LINE:COL)
		
		strikes: 3
	}
}
`;

exports[`Collections Sets disambiguates entries and numeric properties 1`] = `
input: {
	set: Set {
		0 => {
			b: "c"
		}
		1 => {
			bar: "Baz"
		}
		
		1: {
			nah: "Nope"
		}
		foo: -> input.set[1]
		bar: -> input.set.1
	}
}
`;

exports[`Collections Sets indicates references using \`->\` 1`] = `
{
	foo: "Foo"
	bar: {
		name: "Bar"
	}
	baz: Set {
		0 -> {root}.bar
	}
}
`;

exports[`Collections Sets prints basic entries 1`] = `
Set {
	0 => "A"
	1 => "B"
	2 => 204
}
`;

exports[`Collections Sets prints empty sets on one line 1`] = `Set {}`;

exports[`Collections Sets prints empty sets on one line 2`] = `
Set {
	foo: "Bar"
}
`;

exports[`Collections Sets prints nested sets 1`] = `
Set {
	0 => 1
	1 => {
		colour: 65280
		width: 28.52
		height: 10.2
		range: [
			-20
			12
		]
	}
	2 => "3"
	3 => Set {
		0 => 2
		1 => {
			name: "John"
			age: "Older than I look"
			occupation: "Larrikin"
			country: "Australia"
			city: "Melbourne"
		}
		2 => "4"
	}
}
`;

exports[`Collections Sets prints object-type entries 1`] = `
Set {
	0 => "0"
	1 => {
		colour: 65280
		width: 28.52
		height: 10.2
		range: [
			-20
			12
		]
	}
	2 => {
		name: "John"
		age: "Older than I look"
		occupation: "Larrikin"
		country: "Australia"
		city: "Melbourne"
	}
	3 => "C"
}
`;

exports[`Collections Sets prints properties 1`] = `
Set {
	0 => "A"
	1 => "B"
	2 => "C"
	
	name: "Quxabaz"
	customProperty: {
		foo: "Bar"
		baz: "Quux"
	}
}
`;

exports[`Collections Sets prints symbol-type entries 1`] = `
Set {
	0 => @@iterator
	1 => Symbol(Foo)
}
`;

exports[`Collections WeakMaps prints empty maps on one line 1`] = `WeakMap {}`;

exports[`Collections WeakMaps prints properties 1`] = `
WeakMap {
	a: "ABC"
}
`;

exports[`Collections WeakMaps prints properties 2`] = `
WeakMap {
	a: "ABC"
	b: "XYZ"
}
`;

exports[`Collections WeakSets prints empty sets on one line 1`] = `WeakSet {}`;

exports[`Collections WeakSets prints properties 1`] = `
WeakSet {
	a: "ABC"
}
`;

exports[`Collections WeakSets prints properties 2`] = `
WeakSet {
	a: "ABC"
	b: "XYZ"
}
`;
